### ISSUE TITLE
brew update followed by brew upgrade fails for Python's 2to3 conflict (December 2022)

### ISSUE THREAD

**Issue by traversaro** [actions/runner-images#6817]
### Description

If one has on a job:
~~~
brew update
brew upgrade
~~~

the job fails with error:
~~~
 ==> Upgrading python@3.10
  3.10.8 -> 3.10.9 

==> Pouring python@3.10--3.10.9.big_sur.bottle.tar.gz
Error: The `brew link` step did not complete successfully
The formula built, but is not symlinked into /usr/local
Could not symlink bin/2to3
Target /usr/local/bin/2to3
already exists. You may want to remove it:
  rm '/usr/local/bin/2to3'

To force the link and overwrite all conflicting files:
  brew link --overwrite python@3.10

To list all files that would be deleted:
  brew link --overwrite --dry-run python@3.10

Possible conflicting files are:
/usr/local/bin/2to3 -> /Library/Frameworks/Python.framework/Versions/3.11/bin/2to3
/usr/local/bin/idle3 -> /Library/Frameworks/Python.framework/Versions/3.11/bin/idle3
/usr/local/bin/pydoc3 -> /Library/Frameworks/Python.framework/Versions/3.11/bin/pydoc3
/usr/local/bin/python3 -> /Library/Frameworks/Python.framework/Versions/3.11/bin/python3
/usr/local/bin/python3-config -> /Library/Frameworks/Python.framework/Versions/3.11/bin/python3-config
~~~

This is similar to old issues:
* https://github.com/actions/runner-images/issues/2322
* https://github.com/actions/runner-images/issues/4020

### Platforms affected

- [ ] Azure DevOps
- [ ] GitHub Actions - Standard Runners
- [ ] GitHub Actions - Larger Runners

### Runner images affected

- [ ] Ubuntu 18.04
- [ ] Ubuntu 20.04
- [ ] Ubuntu 22.04
- [X] macOS 11
- [X] macOS 12
- [ ] Windows Server 2019
- [ ] Windows Server 2022

### Image version and build link

Image Version: 20221215.1
Link: https://github.com/traversaro/github-actions-brew-update-upgrade-check/actions/runs/3742142787

### Is it regression?

20221211.1

### Expected behavior

The `brew upgrade` should exit fine.

### Actual behavior

The `brew upgrade` fails with the error message in the top of the issue.

### Repro steps

See https://github.com/traversaro/github-actions-brew-update-upgrade-check/blob/2fa847c627c643716ba10d681ea18c86c0b54dec/.github/workflows/brew-update.yml for a MWE .

**Comment by traversaro** [comment]
Possibly related to https://github.com/actions/runner-images/issues/6459 .

**Comment by igorboskovic3** [comment]
Hi @traversaro , thank you, we will take a look.

**Comment by david-engelmann** [comment]
[I'm observing this as well](https://github.com/david-engelmann/turbodbc/pull/5/checks)

**Comment by ryancurrah** [comment]
Also observed this here https://github.com/ryancurrah/lima-and-qemu/actions/runs/3745432855/jobs/6359863993 and here https://github.com/mook-as/lima-and-qemu/actions/runs/3743467832/jobs/6356414606.

**Comment by mikhailkoliada** [comment]
<img width="626" alt="image" src="https://user-images.githubusercontent.com/88318005/208953878-df1ad040-1d59-4c45-a723-722317a687c3.png">

It looks like it the universal python's package used in `actions/setup-python` is installing its symlinks to /usr/local, while I think it should not.

@marko-zivic-93 @dsame @dmitry-shibanov could you please take a look?

**Comment by dsame** [comment]
@Flamefire the workflow does not use `action/setup-python`. It seems the links are created by the macos python package - i see them on the "almost" clean machine with the XCode installed.

**Comment by dsame** [comment]
@traversaro actully it is a normal behaviour. As you see from the log the installation continues normally with just a warning the links were not overwritten. The problem is `brew upgrade` exits with non-zero exit code and `bash` shell runs with `set -e` option. Thus the step signals the error condition despite `brew upgrade` is the very last command or not.

The easiest way to make the step to success is to add `true` after `brew upgrade`:

```
      - run: |
          brew update
          brew upgrade || true
```

**Comment by dsame** [comment]
Hello @fredroy, i see you forces the links by removing them manually in fact it is better to be done with 

`brew link --overwrite python@3.10`

**Comment by maxkratz** [comment]
I'm also observing this behavior (just with running `brew update`): https://github.com/eMoflon/emoflon-ibex-eclipse-build/actions/runs/3764385410/jobs/6398766536

**Comment by Flamefire** [comment]
@dsame 
> @Flamefire the workflow does not use `action/setup-python`. It seems the links are created by the macos python package - i see them on the "almost" clean machine with the XCode installed.

Yes this is what I also observed and wrote in https://github.com/actions/runner-images/issues/6817#issuecomment-1363072043

> As you see from the log the installation continues normally with just a warning the links were not overwritten. The problem is brew upgrade exits with non-zero exit code

We also see this when we install a package with brew (e.g. ccache) which requires Python and hence is automatically upgraded. Ignoring the exit code is not feasible in that case as it might hint to anything besides this issue.

> i see you forces the links by removing them manually in fact it is better to be done with
> 
> `brew link --overwrite python@3.10`

This is also not feasible for us as the exact Python version required for a package to be installed is not known/may change.

Hence our workaround is to remove the symlinks on failure and retry: https://github.com/boostorg/boost-ci/blob/ac5ae7e901e49351d19de278cbf82ce8d2c44216/ci/setup_ccache.sh#L17-L24

@maxkratz 
> I'm also observing this behavior (just with running `brew update` and macOS 12): https://github.com/eMoflon/emoflon-ibex-eclipse-build/actions/runs/3764385410/jobs/6398766536

No, I also made this false assumption. The behavior is **not** triggered by the update (which just downloads meta information similar to `apt-get update`) but the following installation of a package: `brew install p7zip coreutils grep wget curl` (or `brew upgrade` but you don't use that here)

**Comment by maxkratz** [comment]
> @maxkratz
> 
> > I'm also observing this behavior (just with running `brew update` and macOS 12): https://github.com/eMoflon/emoflon-ibex-eclipse-build/actions/runs/3764385410/jobs/6398766536
> 
> No, I also made this false assumption. The behavior is **not** triggered by the update (which just downloads meta information similar to `apt-get update`) but the following installation of a package: `brew install p7zip coreutils grep wget curl` (or `brew upgrade` but you don't use that here)

Thank you for the clarification, you are right.

**Comment by igagis** [comment]
Any chance to have it fixed in the image? All was working before, I wouldn't want to use any workarounds.

**Comment by wwiitt** [comment]
???????????


Thomas Caspersen
________________________________
Fra: Max Kratz ***@***.***>
Sendt: fredag 23. desember 2022 10:05
Til: actions/runner-images ***@***.***>
Kopi: Subscribed ***@***.***>
Emne: Re: [actions/runner-images] brew update followed by brew upgrade fails for Python's 2to3 conflict (December 2022) (Issue #6817)


I'm also observing this behavior (just with running brew update): https://github.com/eMoflon/emoflon-ibex-eclipse-build/actions/runs/3764385410/jobs/6398766536

**Comment by mikhailkoliada** [comment]
@igagis the fix should be made in the `actions/python-versions` repo in first place, it is image - irrelevant.

**Comment by Flamefire** [comment]
> @igagis the fix should be made in the `actions/python-versions` repo in first place, it is image - irrelevant.

This is not true. Many people with this issue do not use that action, the conflicting Python symlink targets e.g `/Library/Frameworks/Python.framework/Versions/3.11/bin/2to3`. See the workflow file from the original issue post: https://github.com/traversaro/github-actions-brew-update-upgrade-check/actions/runs/3742142787/workflow

**Comment by igagis** [comment]
> @igagis the fix should be made in the `actions/python-versions` repo in first place, it is image - irrelevant.

I'm also not using this action, still see the issue

**Comment by mikhailkoliada** [comment]
@Flamefire @igagis python-versions does not depend on setup-python, python-versions provides python which is later used in toolcache used by setup-python, so even if you do not invoke the action itself, you are still affected.

**Comment by igagis** [comment]
I'm no using either of ` setup-python, python-versions ` actions, should I still be affected?

**Comment by mikhailkoliada** [comment]
@igagis yes, because toolcached python is installed during image creation process, no matter will anybody use it or not it is already a part of the image.

**Comment by igagis** [comment]
@mikhailkoliada Ok, do maintainers of `actions/python-versions` know that they need to fix the problem? The `Issues` section is disabled on their github repo.

**Comment by david-engelmann** [comment]
@igagis That seems like a question for @MaksimZhukov
